File: app/region_elections.py
"""EPIC-02 自分に関係する選挙を取得できる。"""
import sqlite3
from datetime import date

from app.regions import get_region


def _matches_region(election: dict, region: dict) -> bool:
    if election["prefecture"] is not None and election["prefecture"] != region["prefecture"]:
        return False
    if election["city"] is not None and election["city"] != region["city"]:
        return False
    return True


def list_elections_for_region(
    conn: sqlite3.Connection, region_id: int, election_type: str | None = None
) -> list[dict]:
    """登録地域（自治体・都道府県）に紐づく実施予定の選挙を取得する。"""
    region = get_region(conn, region_id)
    rows = conn.execute(
        "SELECT * FROM elections WHERE status = 'scheduled' ORDER BY vote_date"
    ).fetchall()
    result = [dict(r) for r in rows if _matches_region(dict(r), region)]
    if election_type is not None:
        result = [e for e in result if e["election_type"] == election_type]
    return result


def next_election_for_region(conn: sqlite3.Connection, region_id: int) -> dict | None:
    """登録地域における次回の選挙を判定する。"""
    today = date.today().isoformat()
    upcoming = [
        e for e in list_elections_for_region(conn, region_id) if e["vote_date"] >= today
    ]
    return upcoming[0] if upcoming else None

File: src/logic/regionElections.js
// EPIC-02 自分に関係する選挙を取得できる。
import { loadState } from "./db.js"
import { getRegion } from "./regions.js"
import { finalizePastElections, listElections } from "./elections.js"

function matchesRegion(election, region) {
  if (election.prefecture !== null && election.prefecture !== region.prefecture) return false
  if (election.city !== null && election.city !== region.city) return false
  return true
}

function today() {
  return new Date().toISOString().slice(0, 10)
}

// ステータスを問わず、登録地域（自治体・都道府県）に紐づく選挙を取得する。
export function listAllElectionsForRegion(regionId) {
  finalizePastElections()
  const state = loadState()
  const region = getRegion(regionId, state)
  return listElections(state).filter((e) => matchesRegion(e, region))
}

// 登録地域（自治体・都道府県）に紐づく実施予定の選挙を取得する。
export function listElectionsForRegion(regionId, electionType = null) {
  let result = listAllElectionsForRegion(regionId).filter((e) => e.status === "scheduled")
  if (electionType !== null) {
    result = result.filter((e) => e.election_type === electionType)
  }
  return result
}

// 登録地域における次回の選挙を判定する。
export function nextElectionForRegion(regionId) {
  const todayStr = today()
  const upcoming = listElectionsForRegion(regionId).filter((e) => e.vote_date >= todayStr)
  return upcoming.length > 0 ? upcoming[0] : null
}
